app/assets/javascripts/listings/ListingDataService.js.coffee

```diff
@@ -0,0 +1,434 @@
+  $http, $localStorage, $q, $state, $translate, $timeout,
+  ExternalTranslateService, ListingConstantsService, ListingEligibilityService, ListingIdentityService,
+  ListingLotteryService, ListingPreferenceService, ListingUnitService, SharedService) ->
+  Service = {}
+  MAINTENANCE_LISTINGS = [] unless MAINTENANCE_LISTINGS
+  Service.listing = {}
+  Service.listings = []
+  Service.openListings = []
+  Service.openMatchListings = []
+  Service.openNotMatchListings = []
+  Service.closedListings = []
+  Service.lotteryResultsListings = []
+  # these get loaded after the listing is loaded
+  Service.AMICharts = []
+  Service.loading = {}
+  Service.error = {}
+  Service.toggleStates = {}
+  Service.listingDownloadURLs = []
+  $localStorage.favorites ?= []
+  Service.favorites = $localStorage.favorites
+  Service.preferenceMap = ListingConstantsService.preferenceMap
+
+  Service.getFavoriteListings = () ->
+    Service.getListingsByIds(Service.favorites, true)
+
+  # Service.checkFavorites makes sure that Service.listings contains our favorited listings
+  # if not, it means the listing doesn't exist and we should remove it from favorites
+  Service.checkFavorites = () ->
+    listing_ids = []
+    Service.listings.forEach (listing) -> listing_ids.push(listing['Id'])
+    Service.favorites.forEach (favorite_id) ->
+      if listing_ids.indexOf(favorite_id) == -1
+        Service.toggleFavoriteListing(favorite_id)
+
+  Service.toggleFavoriteListing = (listing_id) ->
+    # toggle the value for listing_id
+    index = Service.favorites.indexOf(listing_id)
+    if index == -1
+      # add the favorite
+      Service.favorites.push(listing_id)
+    else
+      # remove the favorite
+      Service.favorites.splice(index, 1)
+
+  Service.isFavorited = (listing_id) ->
+    Service.favorites.indexOf(listing_id) > -1
+
+  Service.sortByDate = (sessions) ->
+    # used for sorting Open_Houses and Information_Sessions
+    _.sortBy sessions, (session) ->
+      moment("#{session.Date} #{session.Start_Time}", 'YYYY-MM-DD h:mmA')
+
+  ###################################### Salesforce API Calls ###################################
+
+  Service.getListing = (_id, forceRecache = false, retranslate = false) ->
+    _id = Service.mapSlugToId(_id)
+
+    if Service.listing && Service.listing.Id == _id
+      # return a resolved promise if we already have the listing
+      return $q.when(Service.listing)
+    Service.resetListingData()
+
+    deferred = $q.defer()
+    httpConfig = { etagCache: true }
+    httpConfig.params = { force: true } if forceRecache
+    $http.get("/api/v1/listings/#{_id}.json", httpConfig)
+    .success(
+      Service.getListingResponse(deferred, retranslate)
+    ).cached(
+      Service.getListingResponse(deferred, retranslate)
+    ).error( (data, status, headers, config) ->
+      deferred.reject(data)
+    )
+    return deferred.promise
+
+  # Remove the previous listing and all it's associated data
+  Service.resetListingData = () ->
```

Comment:
looks like this is missing removing units and preferences

Reply:
looks like the original function didn't reset units, preferences, or lottery ranking info. it seems like it would be in line with the intention of this function to add resets of those, but can we think of a reason why those were not already being reset?